tests/integration/ha_tests/helpers.py

```diff
@@ -334,3 +345,150 @@ async def get_units_hostnames(ops_test: OpsTest) -> List[str]:
+    await gather(*awaits)
+
+
+async def get_total_writes(ops_test: OpsTest) -> int:
+    """Gets the total writes from the test application action."""
+    application_name = await get_application_name(ops_test, "application")
+    application_unit = ops_test.model.applications[application_name].units[0]
+    stop_writes_action = await application_unit.run_action("stop-continuous-writes")
+    await stop_writes_action.wait()
+    total_expected_writes = int(stop_writes_action.results["writes"])
+    assert total_expected_writes > 0, "error while getting total writes."
+    return total_expected_writes
+
+
+async def kubectl_delete(ops_test: OpsTest, unit: ops.model.Unit, wait: bool = True) -> None:
+    """Delete the underlying pod for a unit."""
+    kubectl_cmd = (
+        "microk8s",
+        "kubectl",
+        "delete",
+        "pod",
+        f"--wait={wait}",
+        f"-n{ops_test.model_name}",
+        unit.name.replace("/", "-"),
+    )
+    ret_code, _, _ = await ops_test.run(*kubectl_cmd)
+    assert ret_code == 0, "Unit failed to delete"
+
+
+async def insert_focal_to_cluster(ops_test: OpsTest) -> None:
+    """Inserts the Focal Fossa data into the MongoDB cluster via primary replica."""
+    primary = await get_replica_set_primary(ops_test)
+    with await get_direct_mongo_client(ops_test, primary) as client:
+        db = client["new-db"]
+        test_collection = db["test_ubuntu_collection"]
+        test_collection.insert_one({"release_name": "Focal Fossa", "version": 20.04, "LTS": True})
+
+
+async def find_focal_in_cluster(ops_test: OpsTest) -> None:
+    """Checks that all the nodes in the cluster have the Focal Fossa data."""
+    app = await get_application_name(ops_test, APP_NAME)
+    for unit in ops_test.model.applications[app].units:
+        with await get_direct_mongo_client(ops_test, unit.name) as client:
+            db = client["new-db"]
+            test_collection = db["test_ubuntu_collection"]
+            query = test_collection.find({}, {"release_name": 1})
+            release_name = query[0]["release_name"]
+        assert release_name == "Focal Fossa"
+
+
+async def verify_writes(ops_test: OpsTest) -> int:
+    """Verifies that no writes to the cluster were missed.
+
+    Gets the total writes according to the test application and verifies against all nodes
+    """
+    app = await get_application_name(ops_test, APP_NAME)
+    primary = await get_replica_set_primary(ops_test)
+
+    total_expected_writes = await get_total_writes(ops_test)
+    for unit in ops_test.model.applications[app].units:
+        if unit.name == primary:
+            role = "Primary"
+        else:
+            role = "Secondary"
+        with await get_direct_mongo_client(ops_test, unit.name) as client:
+            actual_writes = client[TEST_DB][TEST_COLLECTION].count_documents({})
+        assert (
+            total_expected_writes == actual_writes
+        ), f"{role} {unit.name} missed writes to the db."
```

Comment:
Most of the tests either check a random node and/or a specific one. I think we can recheck all the nodes and do that consistently across tests.